blockchain/vm/internaltx_tracer.go

```diff
@@ -0,0 +1,495 @@
+// prior to the execution of the statement.
+type InternalCall struct {
+	Type    OpCode         `json:"op"`
+	From    common.Address `json:"from"`
+	To      common.Address `json:"to"`
+	Input   string
+	Gas     uint64 `json:"gas"`
+	GasIn   uint64 `json:"gasIn"`
+	GasUsed uint64 `json:"gasUsed"`
+	GasCost uint64 `json:"gasCost"`
+	Value   string `json:"value"`
+	Err     error  `json:"-"`
+	OutOff  *big.Int
+	OutLen  *big.Int
+	Output  []byte
+
+	calls []*InternalCall
+}
+
+// OpName formats the operand name in a human-readable format.
+func (s *InternalCall) OpName() string {
+	return s.Type.String()
+}
+
+// ErrorString formats the tracerLog's error as a string.
+func (s *InternalCall) ErrorString() string {
+	if s.Err != nil {
+		return s.Err.Error()
+	}
+	return ""
+}
+
+// InternalTxTraceResult is returned data after the end of trace-collecting cycle.
+// It implements an object returned by "result" function at call_tracer.js
+type InternalTxTraceResult struct {
+	callType OpCode
+	from     common.Address
+	to       common.Address
+	value    string
+	gas      uint64
+	gasUsed  uint64
+	input    string // hex string
+	output   string // hex string
+	time     time.Duration
+	calls    []*InternalCall
+	error    error
+	reverted revertedInfo
+}
+
+type revertedInfo struct {
+	contract common.Address
+	message  string
+}
+
+// CaptureStart implements the Tracer interface to initialize the tracing operation.
+func (this *InternalTxTracer) CaptureStart(from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int) error {
+	this.ctx["type"] = "CALL"
+	if create {
+		this.ctx["type"] = "CREATE"
+	}
+	this.ctx["from"] = from
+	this.ctx["to"] = to
+	this.ctx["input"] = hexutil.Encode(input)
+	this.ctx["gas"] = gas
+	this.ctx["value"] = value
+
+	return nil
+}
+
+// tracerLog is used to help comparing codes between this and call_tracer.js
+// by following the conventions used in call_tracer.js
+type tracerLog struct {
+	env      *EVM
+	pc       uint64
+	op       OpCode
+	gas      uint64
+	cost     uint64
+	memory   *Memory
+	stack    *Stack
+	contract *Contract
+	depth    int
+	err      error
+}
+
+// CaptureState implements the Tracer interface to trace a single step of VM execution.
+func (this *InternalTxTracer) CaptureState(env *EVM, pc uint64, op OpCode, gas, cost uint64, memory *Memory, logStack *Stack, contract *Contract, depth int, err error) error {
+	if this.err == nil {
+		// Initialize the context if it wasn't done yet
+		if !this.initialized {
+			this.ctx["block"] = env.BlockNumber.Uint64()
+			this.initialized = true
+		}
+		// As this tracer is executed inside the node and there is no reason to stop this
+		// while the node is running, let's skip stop logic from tracers.Tracer
+
+		log := &tracerLog{
+			env, pc, op, gas, cost,
+			memory, logStack, contract, depth, err,
+		}
+		err := this.step(log)
+		if err != nil {
+			this.err = wrapError("step", err)
+		}
```

Comment:
Doesn't this code need to be changed as follows?
If it does, `step` also doesn't need to return anything.
```suggestion
		this.step(log)
		if log.err != nil {
			this.err = wrapError("step", log.err)
		}
```